### PR TITLE
[Feature] artifacts-download-trigger Action

### PR DESCRIPTION
**Pull request by Dexus**
**What**:

Feature: artifacts-download-trigger

This feature reuses code from:
https://github.com/dawidd6/action-download-artifact

and:
https://github.com/Dexus-Forks/workflow-dispatch

**Why**:

Support to trigger workflows and download the resulting artifacts.

**How**:

via tests

**Checklist**:

- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+#        uses: actions/checkout@v2
+#      - name: Download dist
+#        uses: actions/download-artifact@v2
+#        with:
+#          name: build
+#          path: artifacts-download-trigger/dist
+#      - name: Update repo with build
+#        uses: mikeal/publish-to-github-action@master
+#        env:
+#          GITHUB_TOKEN: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:198]:
```suggestion
#          GITHUB_TOKEN: ${{ secrets.BOT_PAT_TOKEN }}

```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          name: artifact
+          path: artifact
+          workflow_conclusion: ''
+          download-artifacts: true
+          force-trigger: true
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
+      - name: Test
+        run: cat artifact/sha | grep $GITHUB_SHA
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:109]:
```suggestion
        run: cat artifact/sha | grep $GITHUB_SHA

```

File: .github/workflows/artifacts-download-trigger-echo1.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+jobs:
+  echo:
+    runs-on: ubuntu-latest
+    environment:
+      name: foo
+      url: www.google.com
+    steps:
+      - name: Echo message
+        run: echo '${{ github.event.inputs.message }}'
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-echo1.yml:19]:
```suggestion
        run: echo '${{ github.event.inputs.message }}'

```

File: .github/workflows/artifacts-download-trigger-echo2.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        description: "Message to echo"
+        required: false
+        default: "this is echo 2"
+
+jobs:
+  echo:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Echo message
+        run: echo '${{ github.event.inputs.message }}'
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-echo2.yml:16]:
```suggestion
        run: echo '${{ github.event.inputs.message }}'

```

File: .github/workflows/artifacts-download-trigger-failing.yml [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+on:
+  workflow_dispatch:
+
+jobs:
+  echo:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Fail
+        run: exit 1
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-failing.yml:11]:
```suggestion
        run: exit 1

```

File: .github/workflows/artifacts-download-trigger-long-running.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+on:
+  workflow_dispatch:
+
+jobs:
+  echo:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Sleep
+        run: sleep 10s
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-long-running.yml:11]:
```suggestion
        run: sleep 10s

```

File: .github/workflows/artifacts-download-trigger-timeout.yml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+on:
+  workflow_dispatch:
+
+jobs:
+  wait:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Sleep
+        run: sleep 1200s
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-timeout.yml:11]:
```suggestion
        run: sleep 1200s

```

File: .github/workflows/artifacts-download-trigger-upload.yml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      - name: Upload first
+        uses: actions/upload-artifact@v2
+        with:
+          name: artifact1
+          path: artifact1
+      - name: Upload second
+        uses: actions/upload-artifact@v2
+        with:
+          name: artifact2
+          path: artifact2
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-upload.yml:43]:
```suggestion
          path: artifact2

```

File: .gitignore [hunk 9]
```diff
@@ -129,3 +129,7 @@ dmypy.json
 
 node_modules/
 .vscode
+
+# Jetbrains
+.idea/
+*.iml
```
Comment (y0urself) [inline .gitignore:135]:
```suggestion
*.iml

```

File: artifacts-download-trigger/action.yml [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    description: 'Should the Artifact downloaded?'
+    required: false
+    default: 'false'
+  force-trigger:
+    description: 'Should the workflow trigger forced?'
+    required: false
+    default: 'false'
+runs:
+  using: node12
+  main: dist/index.js
```
Comment (y0urself) [inline artifacts-download-trigger/action.yml:95]:
```suggestion
  main: dist/index.js

```

File: artifacts-download-trigger/src/debug.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import * as core from '@actions/core'
+
+export function debug(title: string, content: any) {
+  if (core.isDebug()) {
+    core.info(`::group::${title}`)
+    core.debug(JSON.stringify(content, null, 3));
+    core.info('::endgroup::')
+  }
+}
```
Comment (y0urself) [inline artifacts-download-trigger/src/debug.ts:9]:
```suggestion
}

```

File: artifacts-download-trigger/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        core.setFailed(e.toUpperCase());
+      } else if (e instanceof Error) {
+        console.log(e.stack)
+        core.setFailed(e.message);
+      }
+    }
+  }
+}
+
+run()
```
Comment (y0urself) [inline artifacts-download-trigger/src/index.ts:36]:
```suggestion
run()

```

File: artifacts-download-trigger/src/main-dl.ts [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    }
+  } catch (error) {
+    if (typeof error === "string") {
+      core.setFailed(error.toUpperCase()); // works, `e` narrowed to string
+    } else if (error instanceof Error) {
+      core.setFailed(error.message); // works, `e` narrowed to Error
+    }
+    throw error
+  }
+}
```
Comment (y0urself) [inline artifacts-download-trigger/src/main-dl.ts:169]:
```suggestion
}

```

File: artifacts-download-trigger/src/main.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+  } catch (error) {
+    if (typeof error === "string") {
+      core.setFailed(error.toUpperCase());
+    } else if (error instanceof Error) {
+      core.setFailed(error.message);
+    }
+    throw error
+  }
+}
```
Comment (y0urself) [inline artifacts-download-trigger/src/main.ts:106]:
```suggestion
}

```

File: artifacts-download-trigger/src/utils.ts [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+  let hoursStr = hours + '';
+  let minutesStr = minutes + '';
+  let secondsStr = seconds + '';
+
+  if (hours   < 10) {hoursStr   = "0"+hoursStr;}
+  if (minutes < 10) {minutesStr = "0"+minutesStr;}
+  if (seconds < 10) {secondsStr = "0"+secondsStr;}
+  return hoursStr+'h '+minutesStr+'m '+secondsStr+'s';
+}
```
Comment (y0urself) [inline artifacts-download-trigger/src/utils.ts:92]:
```suggestion
}

```

File: artifacts-download-trigger/src/workflow-handler.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    } catch(error) {
+      debug('List workflows error', error);
+      throw error;
+    }
+  }
+
+  private static isFilename(workflowRef: string) {
+    return /.+\.ya?ml$/.test(workflowRef);
+  }
+}
```
Comment (y0urself) [inline artifacts-download-trigger/src/workflow-handler.ts:197]:
```suggestion
}

```

File: artifacts-download-trigger/tsconfig.json [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+  "compilerOptions": {
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019' or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    "strict": true,                           /* Enable all strict type-checking options. */
+    "noImplicitAny": true,                    /* Raise error on expressions and declarations with an implied 'any' type. */
+    "esModuleInterop": true                   /* Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'. */
+  },
+  "exclude": ["node_modules", "**/*.test.ts"]
+}
```
Comment (y0urself) [inline artifacts-download-trigger/tsconfig.json:11]:
```suggestion
}

```

**Review by y0urself — CHANGES_REQUESTED** [review]
We can at least add a newline at the end. :)

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+      - name: Download dist
+        uses: actions/download-artifact@v2
+        with:
+          name: build
+          path: artifacts-download-trigger/dist
+      - name: Invoke echo 1 workflow using this action (do not wait for completion)
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger Message Echo 1
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:51]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

We want to use a token that actually exists ...

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+      - name: Download dist
+        uses: actions/download-artifact@v2
+        with:
+          name: build
+          path: artifacts-download-trigger/dist
+      - name: Invoke echo 2 workflow using this action
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-echo2.yml
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:71]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+        uses: actions/download-artifact@v2
+        with:
+          name: build
+          path: artifacts-download-trigger/dist
+      - name: Invoke 'long-running' workflow and wait for result using this action
+        id: long-running-workflow
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-long-running.yml
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:99]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+        uses: actions/download-artifact@v2
+        with:
+          name: build
+          path: artifacts-download-trigger/dist
+      - name: Invoke 'failing' workflow and wait for result using this action
+        id: failing-workflow
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-failing.yml
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:130]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+        uses: actions/download-artifact@v2
+        with:
+          name: build
+          path: artifacts-download-trigger/dist
+      - name: Invoke 'timeout' workflow and wait for result using this action
+        id: timeout-workflow
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-timeout.yml
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:161]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+        uses: actions/checkout@v2
+      - name: Download
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          name: artifact
+          path: artifact
+          download-artifacts: true
+          force-trigger: true
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:34]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Download
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          name: artifact
+          path: artifact
+          branch: main
+          download-artifacts: true
+          force-trigger: true
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:53]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Download
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          name: artifact
+          path: artifact
+          pr: ${{github.event.pull_request.number}}
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:70]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Download
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          download-artifacts: true
+          force-trigger: true
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:87]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

File: .github/workflows/artifacts-download-trigger-download.yml [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Download
+        uses: ./artifacts-download-trigger
+        with:
+          workflow: artifacts-download-trigger-upload.yml
+          name: artifact
+          path: artifact
+          workflow_conclusion: ''
+          download-artifacts: true
+          force-trigger: true
+          gh_token: ${{ secrets.BOT_PAT_TOKEN }}
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-download.yml:107]:
```suggestion
          gh_token: ${{ secrets.GREENBONE_BOT_TOKEN }}
```

**Review by y0urself — COMMENTED** [review]
We should use our own GreenboneBot token for the testing here?

Personally I would prefer `token` or at least take `gh-token` for the Token Input (so all inputs are uniform with `-`).

Edit: I have seen that all the inputs for the action are mixed up (probably because they are from the two different source repositories). We should stick with `-` I think?

File: .github/workflows/artifacts-download-trigger-build.yml (hunk 1 above)
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:198]: I do not think we will ever need this. We can use the action with the `@v1` tag after merging the code in the correct branch.

File: .github/workflows/artifacts-download-trigger-build.yml [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+      - 'artifacts-download-trigger/**'
+  pull_request:
+    branches: [ main, 'feature/*' ]
+    paths:
+      - 'artifacts-download-trigger/**'
+  workflow_dispatch:
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (y0urself) [inline .github/workflows/artifacts-download-trigger-build.yml:16]:
```suggestion
    runs-on: ubuntu-latest
    name: "Build the artifacts-download-trigger action"
```

File: artifacts-download-trigger/action.yml [hunk 29]
```diff
@@ -0,0 +1,95 @@
+name: Workflow Dispatch and wait to download artifact
```
Comment (y0urself) [inline artifacts-download-trigger/action.yml:1]: I am really not sure if we need all the inputs.
It is a really blowed up inputs list for our purpose, isn't it?

Comment (Dexus) [reply artifacts-download-trigger/action.yml:1]: Better have them then need them!

**Comment by Dexus** [comment]
rebased onto current main branch

File: artifacts-download-trigger/README.md [hunk 30]
```diff
@@ -0,0 +1,25 @@
+# artifacts-download-trigger
+
+This action is used to trigger workflows and download artifacts form the workflow.
```
Comment (bjoernricks) [inline artifacts-download-trigger/README.md:3]:
```suggestion
This action is used to trigger workflows and download artifacts from another workflow.
```

File: artifacts-download-trigger/action.yml [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+    description: Event type
+    required: false
+  run-id:
+    description: Workflow run id
+    required: false
+  run-number:
+    description: Workflow run number
+    required: false
+  name:
+    description: Artifact name (download all artifacts in not specified)
```
Comment (bjoernricks) [inline artifacts-download-trigger/action.yml:37]:
```suggestion
    description: Artifact name (download all artifacts if not specified)
```

File: artifacts-download-trigger/action.yml [hunk 32]
```diff
@@ -0,0 +1,95 @@
...
+    required: false
+  path:
+    description: Where to unpack the artifact
```
Comment (bjoernricks) [inline artifacts-download-trigger/action.yml:40]:
```suggestion
    description: Where to unpack the artifact (default current working directory)
```

File: artifacts-download-trigger/action.yml [hunk 33]
```diff
@@ -0,0 +1,95 @@
...
+  wait-for-completion-timeout:
+    description: 'Maximum amount of time (+unit) to wait to mark workflow as timed out'
+    required: false
+    default: 1h
+  wait-for-completion-interval:
+    description: 'Time to wait (+unit) between two polls to get run status'
+    required: false
+    default: 1m
+  download-artifacts:
+    description: 'Should the Artifact downloaded?'
```
Comment (bjoernricks) [inline artifacts-download-trigger/action.yml:86]:
```suggestion
    description: 'Should the Artifact be downloaded?'
```

File: artifacts-download-trigger/action.yml [hunk 34]
```diff
@@ -0,0 +1,95 @@
...
+    required: false
+    default: 'false'
+  force-trigger:
+    description: 'Should the workflow trigger forced?'
```
Comment (bjoernricks) [inline artifacts-download-trigger/action.yml:90]:
```suggestion
    description: 'Should the workflow trigger be forced?'
```

**Review by bjoernricks — DISMISSED** [review]
artifacts-download-trigger/package-lock.json should not be necessary because we are using yarn and not npm.

**Pull request merged**